### PR TITLE
dex: fold `Position`s into the price indexes

### PR DESCRIPTION
**Pull request by erwanor**
## Describe your changes

This PR:
- fold `Position` records into the DEX price indexes
- remove extra hashing of positions (to compute `Id`s)
- adds a migration script to population the nonconsensus idx

## Checklist before requesting a review

- [x] If this code contains consensus-breaking changes, I have added the "consensus-breaking" label. Otherwise, I declare my belief that there are not consensus-breaking changes, for the following reason:

  > This is consensus breaking, AND requires a migration.

### CODE REVIEW

**Comment by erwanor** [comment]
There's an interesting idea about using the pb type for asset id to short-circuit an expensive field element serialization. But if we do it, that will be in a different PR.

File: crates/core/component/dex/src/component/position_manager.rs [hunk 1]
```diff
@@ -462,28 +463,28 @@ impl<T: StateWrite + ?Sized + Chandelier> PositionManager for T {}
 trait Inner: StateWrite {
     /// Writes a position to the state, updating all necessary indexes.
     ///
-    /// This should be the SOLE ENTRYPOINT for writing positions to the state.
+    /// This should be the **SOLE ENTRYPOINT** for writing positions to the state.
     /// All other position changes exposed by the `PositionManager` should run through here.
     #[instrument(level = "debug", skip_all)]
     async fn update_position(
         &mut self,
+        id: &position::Id,
```
Comment (zbuc) [inline crates/core/component/dex/src/component/position_manager.rs:471]: why is it bad for this fn to re-hash the position ID? it seems like as implemented it is subject to misuse where you could specify the wrong position ID for a position you're updating

Comment (erwanor) [reply crates/core/component/dex/src/component/position_manager.rs:471]: The motivation is that doing this on every single position state update adds up

**Review by zbuc — APPROVED** [review]

**Comment by conorsch** [comment]
Heads up, we've got conflicts with #4567 & #4596 now. I recommend we land #4596 and then work on rebasing this PR to accommodate.

**Comment by conorsch** [comment]
OK, I believe we're unblocked to rebase this one now. Sorry we missed an opportunity to coordinate more closely on starting the Testnet78 migration. @erwanor could you handle the rebase since you know the logic best? Then I'll circle back to test the migration end to end as part of final review.

**Review by conorsch — CHANGES_REQUESTED** [review]
Has conflicts due to Testnet78 migration added recently, oops. Requesting rebase resolution, then will test as part of review.

**Comment by erwanor** [comment]
I have tested this manually, and rebased onto main.

**Comment by conorsch** [comment]
Tested this by setting up a devnet on `v0.77.2`, opening an LP on the pair `penumbra:test_usd`, and confirming I can swap out of it:

```
❯ pcli --home ~/.local/share/pcli-devnet q dex all-positions
Scanning blocks from last sync height 1745 to latest height 1745
[0s] ██████████████████████████████████████████████████       0/0       0/s ETA: 0s
 ID                                                                State    Fee  Sell Price    Reserves
 plpid1xs8dxcp22k7cw9adkqaya5lvvqyv5spwqfemj7kxjam40y6rt2fqms4f8p  opened  0bps   1penumbra  95test_usd
 └──────────────────────────────────────────────────────────────▶          0bps   1test_usd   5penumbra
 ```
 
 Then, after voting for a halt and applying the migration, the position is still visible, and I can still swap into it, seeing the balance updated:
 
 ```
 ❯ pcli --home ~/.local/share/pcli-devnet tx swap 45penumbra --into test_usd
Scanning blocks from last sync height 1911 to latest height 1911
[0s] ██████████████████████████████████████████████████       0/0       0/s ETA: 0s
building transaction [3 actions, 3 proofs]...
finished proving in 0.566 seconds [3 actions, 3 proofs, 2377 bytes]
broadcasting transaction and awaiting confirmation...
transaction broadcast successfully: b75ecd4cfe2a414913d8a698c9e3a16a93d67ab353426deead6e9e68762703dc
transaction confirmed and detected: b75ecd4cfe2a414913d8a698c9e3a16a93d67ab353426deead6e9e68762703dc @ height 1912
Swap submitted and batch confirmed!
You will receive outputs of 45test_usd and 0penumbra. Claiming now...
building transaction [1 actions, 1 proofs]...
finished proving in 0.495 seconds [1 actions, 1 proofs, 585 bytes]
broadcasting transaction and awaiting confirmation...
transaction broadcast successfully: 1163bcceb238b4b805b58123c7cf1beb9a5a46924404f0d04ba77651cdfc4b2a
transaction confirmed and detected: 1163bcceb238b4b805b58123c7cf1beb9a5a46924404f0d04ba77651cdfc4b2a

❯ cargo run -q --release --bin pcli -- --home ~/.local/share/pcli-devnet/ q dex all-positions
Scanning blocks from last sync height 1914 to latest height 1914
[0s] ██████████████████████████████████████████████████       0/0       0/s ETA: 0s
 ID                                                                State    Fee  Sell Price    Reserves
 plpid1xs8dxcp22k7cw9adkqaya5lvvqyv5spwqfemj7kxjam40y6rt2fqms4f8p  opened  0bps   1penumbra  50test_usd
 └──────────────────────────────────────────────────────────────▶          0bps   1test_usd  50penumbra
 ```
 
 That looks right to me! One more migration to write toward 78.

**Review by conorsch — APPROVED** [review]

**Pull request merged**